pkg/attribute/tracker.go

```diff
@@ -99,13 +106,184 @@ func (at *tracker) ApplyRequestAttributes(attrs *mixerpb.Attributes) (*MutableBa
+	// determine the set of attributes to update
+	for _, attrName := range bag.Names() {
+		newValue, _ := bag.Get(attrName)
+		oldValue, _ := mb.Get(attrName)
+
+		if !compareAttributeValue(newValue, oldValue) {
+			mb.values[attrName] = newValue
+			index := at.getWordIndex(attrName)
+
+			switch t := newValue.(type) {
+			case string:
+				output.StringAttributes[index] = t
+			case int64:
+				output.Int64Attributes[index] = t
+			case float64:
+				output.DoubleAttributes[index] = t
+			case bool:
+				output.BoolAttributes[index] = t
+			case time.Time:
+				output.TimestampAttributes[index] = t
+			case time.Duration:
+				output.DurationAttributes[index] = t
+			case []byte:
+				output.BytesAttributes[index] = t
+			case map[string]string:
+				sm := make(map[int32]string, len(t))
+				for smk, smv := range t {
+					sm[at.getWordIndex(smk)] = smv
+				}
+				output.StringMapAttributes[index] = mixerpb.StringMap{Map: sm}
+			}
+		}
+	}
+
+	// if the reverse dictionary was updated, we need to update the protocol dictionary
+	if len(at.currentRevDictionary) != len(at.currentDictionary) {
+		newDict := make(dictionary, len(at.currentRevDictionary))
+		for word, index := range at.currentRevDictionary {
+			newDict[index] = word
+		}
+
+		at.currentDictionary = at.dictionaries.Intern(newDict)
+		output.Dictionary = at.currentDictionary
+	}
+}
+
+func (at *tracker) getWordIndex(newWord string) int32 {
+	if at.currentRevDictionary == nil {
+		// TODO: this should be interned and shared
+		at.currentRevDictionary = make(map[string]int32, len(at.currentDictionary))
+		for index, word := range at.currentDictionary {
+			at.currentRevDictionary[word] = index
+		}
+	}
+
+	index, ok := at.currentRevDictionary[newWord]
+	if !ok {
+		index = int32(len(at.currentRevDictionary))
+		at.currentRevDictionary[newWord] = index
+	}
+
+	return index
+}
+
+func compareAttributeValue(v1, v2 interface{}) bool {
+	switch t1 := v1.(type) {
+	case string:
+		t2, ok := v2.(string)
+		if !ok || t1 != t2 {
+			return false
+		}
+	case int64:
+		t2, ok := v2.(int64)
+		if !ok || t1 != t2 {
+			return false
+		}
+	case float64:
+		t2, ok := v2.(float64)
+		if !ok || t1 != t2 {
+			return false
+		}
+	case bool:
+		t2, ok := v2.(bool)
+		if !ok || t1 != t2 {
+			return false
+		}
+	case time.Time:
+		t2, ok := v2.(time.Time)
+		if !ok || t1 != t2 {
+			return false
+		}
+	case time.Duration:
+		t2, ok := v2.(time.Duration)
+		if !ok || t1 != t2 {
+			return false
+		}
+	case []byte:
+		t2, ok := v2.([]byte)
+		if !ok {
+			return false
+		}
+
+		if len(t1) != len(t2) {
+			return false
+		}
+
+		for i := 0; i < len(t1); i++ {
+			if t1[i] != t2[i] {
+				return false
+			}
+		}
+	case map[string]string:
+		t2, ok := v2.(map[string]string)
+		if !ok {
+			return false
+		}
+
+		if len(t1) != len(t2) {
+			return false
+		}
+
+		for k, v := range t1 {
+			if v != t2[k] {
+				return false
+			}
+		}
+	default:
+		panic(fmt.Sprintf("Unsupported attribute value type: %T", v1))
```

Comment:
is panic really the right thing here?

I feel like we should be avoid using panics wherever possible in this code. Returning an error, which can be passed up and out of ApplyBag() seems feasible.